### ISSUE TITLE
Deserialize fails if `Option<Decimal>` field is explicitly null when using `serde(with)`

### ISSUE THREAD

**Issue by nappa85** [paupino/rust-decimal#675]
Long story short, this works:
```rust
use serde::Deserialize;

#[derive(Debug, Deserialize)]
pub struct A {
    pub a: f64,
    pub b: f64,
    pub c: f64,
}

#[derive(Debug, Deserialize)]
pub struct B {
    pub a: f64,
    pub b: Option<f64>,
}

#[derive(Debug, Deserialize)]
#[serde(untagged)]
pub enum C {
    A(A),
    B(B),
}

fn main() {
    let data = r#"{"a": 1}"#;
    println!("{:?}", serde_json::from_str::<C>(data));
    let data = r#"{"a": 1, "b": null}"#;
    println!("{:?}", serde_json::from_str::<C>(data));
}
```

But if I replace `f64` with `Decimal` and explicitly ask to deserialize from float:
```rust
use serde::Deserialize;
use rust_decimal::Decimal;

#[derive(Debug, Deserialize)]
pub struct A {
    #[serde(with = "rust_decimal::serde::float")]
    pub a: Decimal,
    #[serde(with = "rust_decimal::serde::float")]
    pub b: Decimal,
    #[serde(with = "rust_decimal::serde::float")]
    pub c: Decimal,
}

#[derive(Debug, Deserialize)]
pub struct B {
    #[serde(with = "rust_decimal::serde::float")]
    pub a: Decimal,
    #[serde(with = "rust_decimal::serde::float_option", default)]
    pub b: Option<Decimal>,
}

#[derive(Debug, Deserialize)]
#[serde(untagged)]
pub enum C {
    A(A),
    B(B),
}

fn main() {
    let data = r#"{"a": 1}"#;
    println!("{:?}", serde_json::from_str::<C>(data));
    let data = r#"{"a": 1, "b": null}"#;
    println!("{:?}", serde_json::from_str::<C>(data));
}
```

Second deserialize fails with "data did not match any variant of untagged enum C".

I also replicated removing `Decimal` but using the same deserializer code:
```rust
use serde::Deserialize;
use std::fmt;

#[derive(Debug, Deserialize)]
pub struct A {
    pub a: f64,
    pub b: f64,
    pub c: f64,
}

#[derive(Debug, Deserialize)]
pub struct B {
    pub a: f64,
    #[serde(with = "float_option", default)]
    pub b: Option<f64>,
}

#[derive(Debug, Deserialize)]
#[serde(untagged)]
pub enum C {
    A(A),
    B(B),
}

fn main() {
    let data = r#"{"a": 1}"#;
    println!("{:?}", serde_json::from_str::<C>(data));
    let data = r#"{"a": 1, "b": null}"#;
    println!("{:?}", serde_json::from_str::<C>(data));
}

type Decimal = f64;

pub mod float_option {
    use super::*;
    use serde::Serialize;

    pub fn deserialize<'de, D>(deserializer: D) -> Result<Option<Decimal>, D::Error>
    where
        D: serde::de::Deserializer<'de>,
    {
        deserializer.deserialize_option(OptionDecimalVisitor)
    }

    pub fn serialize<S>(value: &Option<Decimal>, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        match *value {
            Some(ref decimal) => {
                use num_traits::ToPrimitive;
                decimal.to_f64().unwrap().serialize(serializer)
            }
            None => serializer.serialize_none(),
        }
    }
}

struct OptionDecimalVisitor;

impl<'de> serde::de::Visitor<'de> for OptionDecimalVisitor {
    type Value = Option<Decimal>;

    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
        formatter.write_str("a Decimal type representing a fixed-point number")
    }

    fn visit_none<E>(self) -> Result<Option<Decimal>, E>
    where
        E: serde::de::Error,
    {
        Ok(None)
    }

    #[cfg(all(feature = "serde-str", feature = "serde-float"))]
    fn visit_some<D>(self, d: D) -> Result<Option<Decimal>, D::Error>
    where
        D: serde::de::Deserializer<'de>,
    {
        // We've got multiple types that we may see so we need to use any
        d.deserialize_any(DecimalVisitor).map(Some)
    }

    #[cfg(not(all(feature = "serde-str", feature = "serde-float")))]
    fn visit_some<D>(self, d: D) -> Result<Option<Decimal>, D::Error>
    where
        D: serde::de::Deserializer<'de>,
    {
        <Decimal as serde::Deserialize>::deserialize(d).map(Some)
    }
}
```

**Comment by Tony-Samuels** [comment]
See #596, but unfortunately, this is a really annoying limitation of serde that's bitten me multiple times. When using certain tagging representations or using flatten, serde loses the information on what types it's working with.

Serde has refused to implement any proposed solutions, so for now I'm afraid this will just have to be worked around by changing you serialisation format.

I don't expect this will be fixed until a mythical serde 2.0 comes out.

**Comment by nappa85** [comment]
Hoped to not receive this kind of reply, would have been simpler...
Thanks anyway, should we close this issue or leave it open to be easier to be found for other people?
Feel free to close it
